enterprise/lib/codeintel/lsif/protocol/documentation.go

```diff
@@ -0,0 +1,207 @@
+// children "documentationResult" vertices.
+//
+// It allows one represent hierarchical documentation like:
+//
+// "project" (e.g. an HTTP library)
+// -> "documentationResult" (e.g. "HTTP library" library documentation)
+//   -> "documentationResult" (e.g. docs for the "Server" class in the HTTP library)
+//     -> "documentationResult" (e.g. docs for the "Listen" method on the "Server" class)
+//     -> "documentationResult" (e.g. docs for the "Shutdown" method on the "Server" class)
+//       -> ...
+//
+// Note: the "project" -> "documentationResult" attachment above is expressed via a
+// "documentationResult" edge, since the parent is not a "documentationResult" vertex.
+type DocumentationChildrenEdge struct {
+	Edge
+
+	// The parent "documentationResult" vertex ID.
+	InV uint64 `json:"inV"`
+
+	// The ordered children "documentationResult" vertex IDs.
+	OutVs []uint64 `json:"outVs"`
+}
+
+func NewDocumentationChildrenEdge(id, inV uint64, outVs []uint64) DocumentationChildrenEdge {
+	return DocumentationChildrenEdge{
+		Edge: Edge{
+			Element: Element{
+				ID:   id,
+				Type: ElementEdge,
+			},
+			Label: EdgeSourcegraphDocumentationChildren,
+		},
+		OutVs: outVs,
+		InV:   inV,
+	}
+}
+
+// A "documentationResult" vertex
+type DocumentationResult struct {
+	Vertex
+	Result Documentation `json:"result"`
+}
+
+// NewDocumentationResult creates a new "documentationResult" vertex.
+func NewDocumentationResult(id uint64, result Documentation) DocumentationResult {
+	return DocumentationResult{
+		Vertex: Vertex{
+			Element: Element{
+				ID:   id,
+				Type: ElementVertex,
+			},
+			Label: VertexSourcegraphDocumentationResult,
+		},
+		Result: result,
+	}
+}
+
+// A "documentationResult" vertex describes hierarchial project-wide documentation. It represents
+// documentation for a programming construct (variable, function, etc.) or group of programming
+// constructs in a workspace (library, package, crate, module, etc.)
+//
+// The exact structure of the documentation depends on what makes sense for the specific language
+// and concepts being described.
+//
+// Attached to this vertex MUST be two "documentationString" vertices:
+//
+// 1. The {type: "label"} string, a one-line label or this section of documentation.
+// 1. The {type: "detail"} string, a multi-line detailed string for this section of documentation.
+//
+// They can be attached using a "documentationStringEdge".
+type Documentation struct {
+	// A human-readable URL slug identifier for this documentation. It should be unique relative to
+	// sibling Documentation.
+	Slug string `json:"slug"`
+
+	// Whether or not this Documentation is the beginning of a new major section, meaning it and its
+	// its children should be e.g. displayed on their own dedicated page.
+	NewPage bool `json:"newPage"`
+
+	// Tags about the type of content this documentation contains.
+	Tags []DocumentationTag `json:"tags"`
```

Comment:
What do you think about adding detail, title and sort text to `Documentation`?
```suggestion
	Tags []DocumentationTag `json:"tags"`
	
	Label MarkupContent `json:"label"`	
	Detail MarkupContent `json:"detail"`
	
	SortText String `json:"sortText"`// Optional, uses `title` if not defined
```
1:n interactive ranges can be optionally attached to the label/detail via dedicated `documentationLabel`/`documentationDetail` edges. The benefit of inlining these fields is that they enforce a 1:1 relationship between documentation and label/details/sortText. It also makes `Documentation` more self-contained and easier to debug.

Reply:
`sortText` is no longer needed at all, see [this thread](https://github.com/sourcegraph/sourcegraph/pull/20108/files#r620744889).

The reason that the label and detail fields cannot be inlined is because we need to be able to attach a specific range in one of them back to e.g. a hover result (so that hovering over a range in the label or detail text can have a hover result) - this part of your original diagram in specific:

<img width="384" alt="image" src="https://user-images.githubusercontent.com/3173176/116313549-c3e30200-a762-11eb-9add-5e75cc8b0d6d.png">

Now, if your argument is "we could BOTH inline the fields AND have them attached via edges (which can then attach a specific range)" then I guess I'm not sure I would see the point: it would be two separate ways of specifying the same thing, which just adds more complexity.